astropy/modeling/src/projections.c.templ

```diff
@@ -0,0 +1,291 @@
+#include <Python.h> /* Python */
+
+#include <numpy/arrayobject.h> /* Numpy */
+
+#include "prj.h" /* from wcslib */
+#include "wcsmath.h"
+
+
+#if PY_MAJOR_VERSION >= 3
+#define PY3K 1
+#else
+#define PY3K 0
+#endif
+
+
+{% for name, nargs in projections.items() %}
+static PyObject*
+Py_{{ name }}x2s(PyObject* __, PyObject* args, PyObject* kwds)
+{
+  PyObject* x = NULL;
+  PyObject* y = NULL;
+  PyObject* x_array = NULL;
+  PyObject* y_array = NULL;
+  PyObject* phi_array = NULL;
+  PyObject* theta_array = NULL;
+  PyObject* stat_array = NULL;
+  Py_ssize_t n;
+  Py_ssize_t i;
+  PyObject* result = NULL;
+
+  struct prjprm prj;
+
+  int status = -1;
+
+  memset(&prj, 0, sizeof(struct prjprm));
+  prj.phi0 = UNDEFINED;
+  prj.theta0 = UNDEFINED;
+
+  if (!PyArg_ParseTuple(args,
+                        "OO"
+                        {% for i in range(nargs) %}"d" {% endfor %}
+                        ":{{ name }}x2s",
+                        &x, &y
+                        {% for i in range(nargs) %}
+                        {% if loop.first %},{% endif %}
+                        &prj.pv[{{i}} + 1]
+                        {% if not loop.last %},{% endif %}
+                        {% endfor %}
+                        )) {
+    return NULL;
+  }
+
+  /* TODO: This assumes the same shape for the input arrays.  Instead,
+     we should broadcast */
+
+  x_array = PyArray_FromObject(x, PyArray_DOUBLE, 1, NPY_MAXDIMS);
+  if (x_array == NULL) {
+    goto exit;
+  }
+
+  y_array = PyArray_FromObject(y, PyArray_DOUBLE, 1, NPY_MAXDIMS);
+  if (y_array == NULL) {
+    goto exit;
+  }
```

Comment:
For what it's worth, from the modeling package end of things it can be generally assumed that the inputs are already Numpy arrays.  Though I can see how this is useful for making these wrappers generically useful.  I don't know what the overhead is to calling `PyArray_FromObject` on an existing arrays that already has the correct dtype.  I'm guessing it's probably small (since all it takes is like 2 or 3 comparisons).  Just a thought.